app/models/task.rb

```diff
@@ -184,6 +184,30 @@ def days_overdue
     (project.reference_date - task_definition.target_date).to_i / 1.day
   end
 
+  # Action date defines the last time a task has been "actioned", either the
+  # submission date or latest student comment -- whichever is newer
+  def action_date
+    return nil if last_student_comment.nil? || submission_date.nil?
+    return last_student_comment if !last_student_comment.nil? && submission_date.nil?
+    return submission_date      if !submission_date.nil? && last_student_comment.nil?
+    last_student_comment > submission_date ? last_student_comment : submission_date
+  end
+
+  # Returns the last student comment for this task
+  def last_student_comment
+    comments.where(user: project.user).order(:created_at).last
+  end
+
+  # Returns the last tutor comment for this task
+  def last_tutor_comment
+    comments.where(user: project.tutorial.tutor.user).order(:created_at).last
+  end
+
+  # Returns the number of new comments for a user
+  def new_comments_for_user(user)
+    comments.map { |c| c.new_for_user?(user) }.compact.length
```

Comment:
I suggest doing this with a query. Use where to limit tasks to those where recipient_id is user.id, and then count to get the count of these.